python/ray/_private/node.py

```diff
@@ -934,6 +941,120 @@ def _get_cached_port(
 
         return port
 
+    def _wait_for_node_address(self, timeout_s: int = 60) -> str:
+        """Wait until the node_ip_address.json file is avialable.
+
+        node_ip_address.json is created when a ray instance is started.
+
+        Args:
+            timeout_s: If the ip address is not found within this
+                timeout, it will raise ValueError.
+        Returns:
+            The node_ip_address of the current session if it finds it
+            within timeout_s.
+        """
+        for i in range(timeout_s):
+            node_ip_address = self._get_cached_node_ip_address()
+
+            if node_ip_address is not None:
+                return node_ip_address
+
+            time.sleep(1)
+            if i % 10 == 0:
+                logger.info(
+                    "Can't find a `node_ip_address.json` file from "
+                    f"{self.get_session_dir_path()}. "
+                    "Have you started Ray instsance using "
+                    "`ray start` or `ray.init`?"
+                )
+            if i == timeout_s:
+                raise ValueError(
+                    "Can't find a `node_ip_address.json` file from "
+                    f"{self.get_session_dir_path()}. "
+                    f"for {timeout_s} seconds"
+                    "It means the ray instance hasn't started. "
+                    "Did you do `ray start` or `ray.init` on this host?"
+                )
+
+    def _get_cached_node_ip_address(self) -> str:
+        """Get a node address cached on this session.
+
+        If a ray instance is started by `ray start --node-ip-address`,
+        the node ip address is cached to a file node_ip_address.json.
+        Otherwise, the file exists, but it is emptyl.
+
+        This API is process-safe, meaning the file access is protected by
+        a file lock.
+
+        Returns:
+            node_ip_address cached on the current node. None if the node
+            the file doesn't exist, meaning ray instance hasn't been
+            started on a current node. If node_ip_address is not written
+            to a file, it means --node-ip-address is not given, and in this
+            case, we find the IP address ourselves.
```

Comment:
Unfortunately, it doesn't work well with one unit test (there's a test that has 2 different docker container on the same host, which means it will have 2 raylets with the same session_dir with different node ip address).

Reply:
Maybe we can make those containers use the same IP address, but I feel like the current way is probably safer to merge last minute

Reply:
Since two raylets share the same session_dir, writing to a file doesn't work very well...